src/translator/html.cpp

```diff
@@ -622,7 +761,14 @@ void HTML::hardAlignments(Response const &response, std::vector<std::vector<size
         float currScore = response.alignments[sentenceIdx][t][currSentenceIdx];
         float prevScore = response.alignments[sentenceIdx][t - 1][prevSentenceIdx];
 
-        if (currScore >= prevScore) {
+        TagStack const &currTagStack = sourceTokenSpans[offset + 1 + currSentenceIdx]->tags;
+        TagStack const &prevTagStack = sourceTokenSpans[offset + 1 + prevSentenceIdx]->tags;
+
+        // If this token has more markup, or a better score than the previous
+        // token (and they together are part of a word-ish thing) then mark
+        // this word as aligning. Otherwise just copy the alignment source of
+        // the previous token.
+        if (isSubset(prevTagStack, currTagStack) || currScore >= prevScore) {
```

Comment:
Would appreciate some comments on the complexity of the overall process and relevant areas. This comparison is O(length), the tag comparisons can go full document length?

Reply:
Before we start: my complexity theory knowledge is … lacking. Please correct me if I'm wrong.

`isSubset` (now `extends`) is O(N), where N is the stack depth, but that stack pretty small on average. Maybe we should have a limit on how deep you can nest tags in the parser, just to avoid denial of service attacks.

This whole section I think is worst case O(N*N), where N in number of subwords in the document. Worst case is all subwords form a single word together and alignment scores keep getting higher as you go further down the document. On average again, the are only limited subwords per word so O(N).

The expensive bit is a bit higher where it looks for straggers, in `HTML::restoreTarget()`. That bit is O(N*N) where N is the number of spans (i.e. bits between tags) and that bit can use optimisations. A simple hashset would be sufficient there to get it back to O(N).

But putting this all in context, I've profiled various pages, both text heavy and tag heavy, and I've never seen the HTML parser even pop up in the top samples of the profiler. It's always either the model loading code (specifically `marian::binary::loadItems`) or the graph code that does the actual computations.